### DISCUSSION TITLE
How do supported functionalities of FMU effect the training of NeuralFMU?

### DISCUSSION THREAD

**Discussion by awecefil** [ThummeTo/FMIFlux.jl#143] [Q&A]
Currently, I use FMIFlux.jl to combine FMU and NN to do parameter calibration / learning unmodeled effects. 

To export FMU from simulation model, I use OpenModelica instead of Dymola because it is an open source. However, after using `fmiInfo()` to show FMU information, I found that OpenModelica seems not support some functionalities of FMU which may effect the FMIFlux.jl I guess.

Here is the FMU information I use right now (_the FMU is exported by OpenModelica_)：

```
#################### Begin information for FMU ####################
        Model name:                     SpringPendulumExtForce1D
        FMI-Version:                    2.0
        GUID:                           {b0415239-ce28-4aef-83b7-52e90d8a4aed}
        Generation tool:                OpenModelica Compiler OpenModelica v1.23.0 (64-bit)
        Generation time:                2024-08-02T16:41:03Z
        Var. naming conv.:              structured
        Event indicators:               0
        Inputs:                         1
                10 ["force.f", "extForce"]
        Outputs:                        2
                6 ["a", "accSensor.a"]
                19 ["v", "speedSensor.v"]
        States:                         2
                0 ["mass.stateSelect", "force.useSupport", "mass.s"]
                1 ["mass.v"]
        Supports Co-Simulation:         true
                Model identifier:       SpringPendulumExtForce1D
                Get/Set State:          true
                Serialize State:        true
                Dir. Derivatives:       true
                Var. com. steps:        true
                Input interpol.:        true
                Max order out. der.:    1
        Supports Model-Exchange:        true
                Model identifier:       SpringPendulumExtForce1D
                Get/Set State:          false
                Serialize State:        false
                Dir. Derivatives:       true
##################### End information for FMU #####################
```

### What I want to know :

1. For CS-type FMU, does all these functionalities(`Get/Set State`, `Serialize State`,...etc) are needed to make sure there are no problems when training NeuralFMU with FMIFlux.jl? 
2. For ME-type FMU, the `Get/Set State` and `Serialize State` are false. Will they make the training of NeuralFMU with FMIFlux.jl go wrong? Because I really encounter some issues when using ME-type NeuralFMU.

By the way, I have known that Directional Derivatives should be true because the Jocobian of FMU is needed to keep the NeuralFMU trainable (from [https://doi.org/10.3384/ecp21181297](https://doi.org/10.3384/ecp21181297)). To people who want to enable this functionality, you can refer to this [github issue](https://github.com/OpenModelica/OpenModelica/issues/12749)

**Comment by ThummeTo** [comment]
Hi,
for now, two things are important (both are needed!):
- fmiGet/SetState: if you have a discontinuous system, because we need to sample sensitivities at state changes (at least). Fallback is open research (e.g. OpenSCALING project), for some systems we can ignore the sensitivities over the state-jump and still have a "good-enough" gradient, but this can be dangerous...
- fmiGetDirectionalDerivatives: for well implemented FMUs, this is the fastest way to get sensitivities (so we don't need to sample them). Fallback is we do central differences based on samples drawn from the FMU.

**Reply by ThummeTo** [reply]
I personally don't work with CS very much (because ME offers more potential to be combined with ANNs). Above applies for ME FMUs.

For CS-FMUs, I need to think about first... I am not sure if there is a way to determine event sensitivities correctly for CS-FMUs...
